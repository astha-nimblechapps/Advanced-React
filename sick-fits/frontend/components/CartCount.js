import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { TransitionGroup, CSSTransition } from "react-transition-group";

const Dot = styled.div`
  background: ${props => props.theme.black};
  color: white;
  border-radius: 50%;
  padding: 0.5rem;
  line-height: 2.5rem;
  min-width: 3.5rem;
  margin-left: 1rem;
  font-weight: 100;
  font-feature-settings: "tnum";
  font-variant-numeric: tabular-nums;
  text-align: center;
`;

const AnimatedStyles = styled.span`
  position: relative;
  .count {
    display: block;
    position: relative;
    transition: all 0.4s;
    backface-visibility: hidden;
  }
  .count-enter {
    transform: scale(4) rotateX(0.5turn);
  }
  .count-enter-active {
    transform: rotateX(0);
  }
  .count-exit {
    top: 0;
    position: absolute;
    transform: rotateX(0);
  }
  .count-exit-active {
    transform: scale(4) rotateX(0.5turn);
  }
`;

const CartCount = props => (
  <AnimatedStyles>
    <TransitionGroup>
      <CSSTransition
        unmountOnExit
        className="count"
        classNames="count"
        key={props.count}
        timeout={{ enter: 400, exit: 400 }}
      >
        <Dot>{props.count}</Dot>
      </CSSTransition>
    </TransitionGroup>
  </AnimatedStyles>
);

export default CartCount;
