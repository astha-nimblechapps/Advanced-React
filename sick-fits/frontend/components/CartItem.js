import React from "react";
import formatMoney from "../lib/formatMoney";
import styled from "styled-components";
import PropTypes from "prop-types";
import RemoveFromCart from "./RemoveFromCart";
import { Typography, Label } from "./../lib/exim-component";

const CartItemStyle = styled.div`
  padding: 1rem 0;
  border-bottom: 1px solid ${props => props.theme.lightgrey};
  display: grid;
  align-items: center;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 10px;
  }
  h3,
  p {
    margin: 0px;
  }
`;

const CartItem = props => {
  if (!props.cartItem.item)
    return (
      <CartItemStyle>
        <p>This item has been removed</p>{" "}
        <RemoveFromCart id={props.cartItem.id} />
      </CartItemStyle>
    );
  return (
    <CartItemStyle>
      <img
        src={props.cartItem.item.image}
        alt={props.cartItem.item.title}
        width="100"
      />
      <div className="cart-item-details">
        <Typography useFor="subtitle">{props.cartItem.item.title}</Typography>
        <p>
          {props.cartItem.color}
          {props.cartItem.size}
        </p>
        <p>
          <Label color="green">
            {formatMoney(props.cartItem.item.price * props.cartItem.quantity)}
          </Label>
          {" - "}
          <Label color="lightNeutral">
            <em>
              {props.cartItem.quantity} &times;{" "}
              {formatMoney(props.cartItem.item.price)}
            </em>
          </Label>
        </p>
      </div>
      <RemoveFromCart id={props.cartItem.id} />
    </CartItemStyle>
  );
};

CartItem.propTypes = {
  cartItem: PropTypes.object.isRequired
};

export default CartItem;
