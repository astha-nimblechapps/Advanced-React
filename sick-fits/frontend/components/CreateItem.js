import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Router from 'next/router';
import Form from './styles/Form';
import formatMoney from '../lib/formatMoney';
import Error, {ErrorStyles} from './ErrorMessage';
import { Label, Input, Avatar, Button, Card } from '../lib/exim-component';

const ADD_ITEMS = gql`
  mutation ADD_ITEMS(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`;

class CreateItem extends Component {
  state = {
    title: '',
    description: '',
    image: '',
    largeImage: '',
    price: 0,
    errorMessage:'',
   
  };
  handleChange = e => {
    //console.log(e.target.value)
    const { name, type, value } = e.target;
    const val = type === 'number' ? parseFloat(value) : value;
    this.setState({ [name]: val });
  };

  uploadFile = async e => {
    const files = e.target.files;
    const data = new FormData();
    data.append('file', files[0]);
    data.append('upload_preset', 'demoastha');
    //https://res.cloudinary.com/dvo5mxjiw/image/upload/v1542695244/sample.jpg
    const res = await fetch('https://api.cloudinary.com/v1_1/dvo5mxjiw/image/upload', {
      method: 'POST',
      body: data,
    });
    const file = await res.json();
    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url,
    });
  };

  renderStatus = async (e, createItem) => {
    e.preventDefault();
    if (!this.state.title || !this.state.price || !this.state.description || !this.state.image || !this.state.largeImage) {
      this.setState({ errorMessage: "All Fields are required.!!" });
    } else {
      const res = await createItem();              
             // console.log(res);
              Router.push({
                pathname: '/item',
                query: { id: res.data.createItem.id },
              });
      this.setState({ errorMessage:''});
    }
  };

  render() {
    return (
      <Mutation mutation={ADD_ITEMS} variables={this.state}>
        {(createItem, { loading, error }) => (
          <Card>
          <Form
            data-test="form"
            onSubmit={async e => {
             this.renderStatus(e,createItem)
            }}
          >
            {this.state.errorMessage ? (
                      <ErrorStyles>
                        <p data-test="graphql-error">
                          <strong>Shoot!</strong>
                          {this.state.errorMessage}
                        </p>
                      </ErrorStyles>
                    ) : (
                      <Error error={error} />
                    )}
               
            <fieldset disabled={loading} aria-busy={loading}>
            <h6 style={{marginTop: '8px', marginBottom: '8px'}}>Add Item</h6>
            <Label size="small">Image:</Label>
              
                <Input
                  type="file"
                  id="file"
                  name="file"
                  placeholder="Upload an image"
                  required
                  onChange={this.uploadFile}
                />
                {this.state.image && (
                  <Avatar src={ this.state.image}></Avatar>
                  // <img width="100" src={this.state.image} alt="Upload Preview" />
                )}
             
             <Label size="small"> Title:</Label>
               
                <Input
                  type="text"
                  id="title"
                  name="title"
                  placeholder="Title"
                  required
                  value={this.state.title}
                  onChange={this.handleChange}
                />
           

             <Label size="small">Price:</Label>
              
                <Input
                  type="number"
                  id="price"
                  name="price"
                  placeholder="Price"
                  required
                  value={this.state.price}
                  onChange={this.handleChange}
                />
            
              {/* <label htmlFor="color">
                Select Color
                    <select 
                    name="color"
                    value={this.state.color} 
                    onChange={this.handleChange} >
                    <option value="none">Please Select Color</option>
                  <option value="Red">Red</option>
                    <option value="Black">Black</option>
                    <option value="White">White</option>
                  </select>
              </label>
              <label htmlFor="size">
                Select Size
                    <select 
                    name="size"
                    value={this.state.size} 
                    onChange={this.handleChange} >
                    <option value="none">Please Select Size</option>
                  <option value="22">22</option>
                    <option value="23">23</option>
                    <option value="24">24</option>
                  </select>
              </label> */}
           <Label size="small">Description:</Label>
                
                <Input
                type="text"
                  id="description"
                  name="description"
                  placeholder="Enter A Description"
                  required
                  value={this.state.description}
                  onChange={this.handleChange}
                />
           
              <Button type="submit" loader={loading}>Submit</Button>
            </fieldset>
          </Form>
          </Card>
        )}
      </Mutation>
    );
  }
}

export default CreateItem;
export { ADD_ITEMS };
