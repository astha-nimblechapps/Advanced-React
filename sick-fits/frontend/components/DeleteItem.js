import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { ALL_ITEMS_QUERY } from './Items';
import styled from 'styled-components';
import { GET_ALL_ITEMS } from './Items'
const DeleteBtn = styled.button`
    width: 100%;
    border: 1px solid;
    text-align: center;
    font-family: none;
`;

const DELETE_ITEMS = gql`
  mutation DELETE_ITEMS($id: ID!) {
    deleteItem(id: $id) {
      id
    }
  }
`;

class DeleteItem extends Component {
  update = (cache, payload) => {
    const data = cache.readQuery({ query: GET_ALL_ITEMS });
    console.log(data, payload);
    data.items = data.items.filter(item => item.id !== payload.data.deleteItem.id);
    cache.writeQuery({ query: GET_ALL_ITEMS, data });
  };
  render() {
    return (
      <Mutation
        mutation={DELETE_ITEMS}
        variables={{ id: this.props.id }}
        update={this.update}
      >
        {(deleteItem, { error }) => (
          <DeleteBtn
            onClick={() => {
              if (confirm('Are you sure you want to delete this item?')) {
                deleteItem().catch(err => {
                  alert(err.message);
                });
              }
            }}
          >
            {this.props.children}
          </DeleteBtn>
        )}
      </Mutation>
    );
  }
}

export default DeleteItem;
