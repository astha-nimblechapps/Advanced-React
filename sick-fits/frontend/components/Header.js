import Nav from './Nav'
import styled from 'styled-components';
import Link from 'next/link';

const Logo = styled.div`
    font-size: 2.3rem;
    margin-left: 2rem;
    margin-bottom: 2px;
    position: relative;
    z-index: 2;
    transform: skew(-7deg);
    a{
        padding: 0.5rem 1rem;
        background: ${ props => props.theme.black };
        color: white;
        text-decoration: none;
        text-transform:uppercase;
    }
    @media (max-width: 1300px) {
        margin: 0px;
        text-align: center;
    }
`;

const StyledHeader = styled.header`
  .bar {
    border-bottom: 3px solid ${props => props.theme.black};
    display: grid;
    grid-template-columns: auto 1fr;
    justify-content: space-between;
    align-items: stretch;
    @media (max-width: 1300px) {
      grid-template-columns: 1fr;
      justify-content: center;
    }
  }
  .sub-bar {
    display: grid;
    grid-template-columns: 1fr auto;
    border-bottom: 1px solid ${props => props.theme.lightgrey};
  }
`;

const Header = () =>(
    <StyledHeader>
        <div className="bar ">
            <Logo>
                <Link href="/"> 
                    <a>
                        Demo
                    </a> 
                </Link>
            </Logo>
            <Nav/>
        </div>
        <div className="sub-bar">
            <p> Search Here
            </p>
        </div>
    </StyledHeader>
);

export default Header;