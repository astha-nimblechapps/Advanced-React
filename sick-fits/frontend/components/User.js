import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';

const LOGGED_USER = gql`
  query {
    me {
      id
      email
      name
      permissions
    }
  }
`;

const User = props => (
  <Query {...props} query={LOGGED_USER}>
    {payload => props.children(payload)}
  </Query>
);

User.PropTypes = {
  children: PropTypes.func.isRequired,
};

export default User;
export { LOGGED_USER };
