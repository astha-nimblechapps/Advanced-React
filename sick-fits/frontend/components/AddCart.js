import React from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";
import styled from "styled-components";
import User, { LOGGED_USER } from "./User";

const ADDCART = styled.button`
  width: 30%;
  /* border: 1px solid; */
  text-align: center;
  font-family: none;
  border: none;
`;

const ADD_TO_CART = gql`
  mutation ADD_TO_CART($id: ID!) {
    addToCart(id: $id) {
      id
      quantity
    }
  }
`;

const ADD_TO_TEMP_CART = gql`
  mutation ADD_TO_TEMP_CART(
    $id: ID!
    $token: String!
    $color: String
    $size: String
  ) {
    addToTempCart(id: $id, token: $token, color: $color, size: $size) {
      id
      quantity
      color
      size
    }
  }
`;
/**
 * e.preventDefault();
    const res = await updateItemMutation({
      variables: {
        id: this.props.id,
        ...this.state
      }
    });
    console.log("Updated..!!!");
 */

class AddCart extends React.Component {
  addTempCart = async (e, addTempCart) => {
    e.preventDefault();
    if (this.props.color === '' && this.props.size === '' && this.props.color === 'none' && this.props.size === 'none') {
      alert("Please Select Color and Size both")
    } else {
      if (localStorage.getItem("randomId")) {
        console.log("true");
        await addTempCart({
          variables: {
            id: this.props.id,
            token: localStorage.getItem("randomId"),
            color: this.props.color,
            size: this.props.size
          }
        });
      } else {
        console.log("false");
        const min = 1;
        const max = 100;
        const rand = min + Math.random() * (max - min);
        // console.log(rand)
        this.setState({ token: rand });
        localStorage.setItem("randomId", rand);
        // console.log(JSON.stringify(localStorage.getItem("randomId")))
        await addTempCart({
          variables: {
            id: this.props.id,
            token: localStorage.getItem("randomId"),
            color: this.props.color,
            size: this.props.size
          }
        });
      }
    }
  };

  render() {
    const id = this.props.id;

    return (
      <User>
        {({ data: { me } }) => {
          if (me)
            return (
              <Mutation
                mutation={ADD_TO_CART}
                variables={{
                  id
                }}
                refetchQueries={[{ query: LOGGED_USER }]}
              >
                {(addToCart, { loading }) => (
                  <ADDCART disabled={loading} onClick={addToCart}>
                    Add{loading && "ing"}To Cart
                  </ADDCART>
                )}
              </Mutation>
            );
          else {
            return (
              <Mutation mutation={ADD_TO_TEMP_CART}>
                {(addToTempCart, { loading }) => (
                  <ADDCART
                    disabled={loading}
                    onClick={e => this.addTempCart(e, addToTempCart)}
                  >
                    Add{loading && "ing"}To Cart
                  </ADDCART>
                )}
              </Mutation>
            );
          }
        }}
      </User>
    );
  }
}

export default AddCart;
export { ADD_TO_CART, ADDCART };
