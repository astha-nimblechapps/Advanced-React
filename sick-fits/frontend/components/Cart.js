import React from "react";
import { Query, Mutation } from "react-apollo";
import gql from "graphql-tag";
import { adopt } from "react-adopt";
import CloseButton from "./styles/CloseButton";
import Supreme from "./styles/Supreme";
import SickButton from "./styles/SickButton";
import CartStyles from "./styles/CartStyles";
import User from "./User";
import CartItem from "./CartItem";
import calcTotalPrice from "../lib/calcTotalPrice";
import formatMoney from "../lib/formatMoney";
import Payment from "./Payment";
import Router from 'next/router';

const LOCAL_STATE = gql`
  query {
    cartOpen @client
  }
`;

const TOOGLE_CART = gql`
  mutation {
    toogleCart @client
  }
`;

const TEMP_DATA = gql`
  query($token: String) {
    tempCartItem(token: $token) {
      id
      color
      size
      quantity
      item {
        image
        id
        price
        description
        title
      }
    }
  }
`;



class Cart extends React.Component {
  
  
  constructor(props) {
    super(props);
    this.state = {
      tokenId: ""
    }
  }


  componentDidMount() {
    if(localStorage.getItem("randomId")){
      this.setState({
        tokenId: localStorage.getItem("randomId").toString()
      })
    }
    // this.setState({
    //   randomUser: localStorage.getItem("randomId")
    // });
  }

  goToLogin = async(toogleCart) =>{
    await toogleCart();
    Router.push({
      pathname:'/signin'
    })
  }
  
  render() {
    const Composed = adopt({
      user: ({ render }) => <User>{render}</User>,
      toogleCart: ({ render }) => (
        <Mutation mutation={TOOGLE_CART}>{render}</Mutation>
      ),
      localState: ({ render }) => <Query query={LOCAL_STATE}>{render}</Query>,
      tempCart: ({ render }) => <Query query={TEMP_DATA} variables={{ token : this.state.tokenId}}>{render}</Query>
    });
    return (
      <Composed>
        {({ user, toogleCart, localState, tempCart }) => {
          const me = user.data.me;
          const cart = tempCart.data.tempCartItem;
          console.log(cart)
          if (!me){ 
            if(!cart) return <p> No item</p>
            return ( 
            <CartStyles open={localState.data.cartOpen}>
            <header>
              <CloseButton onClick={toogleCart} title="close">
                &times;
              </CloseButton>
              <Supreme>Your Cart</Supreme>
              <p>
                You have {cart.length} itme
                {cart.length === 1 ? "" : "s"} in your cart
              </p>
            </header>
            <ul>
              {cart.map(cartItem => (
              //  console.log(cartItem)
                <CartItem key={cartItem.id} cartItem={cartItem} />
              ))}
            </ul>
            <footer>
              <p>{formatMoney(calcTotalPrice(cart))}</p>
           
              {
                cart.length ? <Payment cart={cart}>
                <SickButton onClick={() => this.goToLogin(toogleCart)}>Checkout</SickButton>
              </Payment> : null
              }
              {/* {cart.length && (
                <Payment>
                  <SickButton>Checkout</SickButton>
                </Payment>
              )} */}
            </footer>
            </CartStyles>
          )
        }
          return (
            <CartStyles open={localState.data.cartOpen}>
              <header>
                <CloseButton onClick={toogleCart} title="close">
                  &times;
                </CloseButton>
                <Supreme>Your Cart</Supreme>
                <p>
                  You have {me.cart.length} itme
                  {me.cart.length === 1 ? "" : "s"} in your cart
                </p>
              </header>
              <ul>
                {me.cart.map(cartItem => (
                  <CartItem key={cartItem.id} cartItem={cartItem} />
                ))}
              </ul>
              <footer>
                <p>{formatMoney(calcTotalPrice(me.cart))}</p>
                {me.cart.length && (
                  <Payment>
                    <SickButton>Checkout</SickButton>
                  </Payment>
                )}
              </footer>
            </CartStyles>
          );
        }}
      </Composed>
    );
  }
}

export default Cart;
export { TOOGLE_CART, LOCAL_STATE, TEMP_DATA };
