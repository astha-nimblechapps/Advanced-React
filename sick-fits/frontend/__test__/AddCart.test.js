import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import { ApolloConsumer } from 'react-apollo';
import AddCart, { ADD_TO_CART } from '../components/AddCart';
import { LOGGED_USER } from '../components/User';
import { fakeUser, fakeCartItem } from '../lib/testUtils';


const mocks = [
    {
      request: { query: LOGGED_USER },
      result: {
        data: {
          me: {
            ...fakeUser(),
            cart: [],
          },
        },
      },
    },
    {
      request: { query: LOGGED_USER },
      result: {
        data: {
          me: {
            ...fakeUser(),
            cart: [fakeCartItem()],
          },
        },
      },
    },
    {
      request: { query: ADD_TO_CART, variables: { id: 'abc' } },
      result: {
        data: {
          addToCart: {
            ...fakeCartItem(),
            quantity: 1,
          },
        },
      },
    },
  ];


  describe('<AddToCart/>', () => {
    it('renders and matches', async () => {
      const wrapper = mount(
        <MockedProvider mocks={mocks}>
          <AddCart id="abc" />
        </MockedProvider>
      );
      await wait();
      wrapper.update();
      expect(toJSON(wrapper.find('button'))).toMatchSnapshot();
    });
  
    it('adds item in cart', async () => {
        let apolloClient;
        const wrapper = mount(
          <MockedProvider mocks={mocks}>
            <ApolloConsumer>
              {client => {
                apolloClient = client;
                return <AddCart  id="abc" />;
              }}
            </ApolloConsumer>
          </MockedProvider>
        );
        await wait();
        wrapper.update();
        const { data: { me } } = await apolloClient.query({ query: LOGGED_USER });
        // console.log(me);
        expect(me.cart).toHaveLength(0);
  
        wrapper.find('button').simulate('click');
        await wait();
    
        const { data: { me: obj } } = await apolloClient.query({ query: LOGGED_USER });
        //console.log(obj);
        expect(obj.cart).toHaveLength(1);
        expect(obj.cart[0].id).toBe('omg123');
        expect(obj.cart[0].quantity).toBe(3);
      });

      it('change add to adding', async () => {
        const wrapper = mount(
          <MockedProvider mocks={mocks}>
            <AddCart id="abc" />
          </MockedProvider>
        );
        await wait();
        wrapper.update();
        //console.log(wrapper.debug())
        expect(wrapper.text()).toContain('AddTo Cart');
        wrapper.find('button').simulate('click');
        expect(wrapper.text()).toContain('AddingTo Cart');
      });
    
  });
  