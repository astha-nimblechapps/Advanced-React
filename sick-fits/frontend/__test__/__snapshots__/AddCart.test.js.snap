// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddToCart/> renders and matches 1`] = `
<button
  className="AddCart__ADDCART-sc-12cdt5i-0 bMsvre"
  disabled={false}
  onClick={[Function]}
>
  Add
  To Cart
</button>
`;
